qiskit/algorithms/phase_estimators/iphase_estimation.py

```diff
@@ -0,0 +1,193 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+
+"""The Iterative Quantum Phase Estimation Algorithm."""
+
+
+from typing import Optional, Union
+import numpy
+import qiskit
+from qiskit.circuit import QuantumCircuit, QuantumRegister
+from qiskit.circuit.classicalregister import ClassicalRegister
+from qiskit.providers import BaseBackend, Backend
+from qiskit.utils import QuantumInstance
+from qiskit.algorithms.algorithm_result import AlgorithmResult
+
+
+class IPhaseEstimation:
+    """Run the Iterative quantum phase estimation (QPE) algorithm.
+
+    Given a unitary circuit and a circuit preparing an eigenstate, return the phase of the
+    eigenvalue as a number in :math:`[0,1)` using the iterative phase estimation algorithm.
+
+    [1]: Dobsicek et al. (2006), Arbitrary accuracy iterative phase estimation algorithm as a two
+       qubit benchmark, `arxiv/quant-ph/0610214 <https://arxiv.org/abs/quant-ph/0610214>`_
+    """
+
+    def __init__(self,
+                 quantum_instance: Optional[Union[QuantumInstance, BaseBackend]] = None) -> None:
+
+        """Args:
+            quantum_instance: The quantum instance on which the circuit will be run.
+        """
+        if isinstance(quantum_instance, (Backend, BaseBackend)):
+            quantum_instance = QuantumInstance(quantum_instance)
+        self._quantum_instance = quantum_instance
+
+    def construct_circuit(self,
+                          problem: dict,
```

Comment:
For consistency with `PhaseEstimation` can we explicitly pass the unitary etc instead of a problem dict? If we want to use an object to wrap the problem information I think we should introduce an analogous class to the `EstimationProblem` in amplitude estimation that is also used in the other phase estimation algorithms.

Reply:
done